src/jit/flowgraph.cpp

```diff
@@ -1621,15 +1694,152 @@ void Compiler::fgReplacePred(BasicBlock* block, BasicBlock* oldPred, BasicBlock*
     noway_assert(newPred != nullptr);
     assert(!fgCheapPredsValid);
 
-    flowList* pred;
+    // Make sure the list is sorted before manipulating.
+    bool sorted = fgExpensiveSortedPredCheck(block);
 
-    for (pred = block->bbPreds; pred != nullptr; pred = pred->flNext)
+#if DEBUG
+    // Also make sure that our refCount is correct before modifying.
+    // If it is not, keep track of that so that we know replacePreds is
+    // not the problem.
+
+    bool refCountCorrect = true;
+
+    if (JitConfig.JitSlowDebugChecksEnabled() != 0)
     {
-        if (oldPred == pred->flBlock)
+        unsigned refCount = 0;
+
+        // Make sure the ref count to this block is still correct.
+        for (flowList* pred = block->bbPreds; pred; pred = pred->flNext)
         {
-            pred->flBlock = newPred;
+            refCount += pred->flDupCount;
+        }
+
+        if (refCount != block->bbRefs)
+        {
+            refCountCorrect = false;
+        }
+    }
+
+#endif // DEBUG
+
+    flowList** elementBeforeMovingPred;
```

Comment:
nit: I wouldn't call this "elementBeforeMovingPred" because "element" might be construed to be a predecessor, which it is not. I would call it "ptrToPred", or "ptrToMovingPred".